### ISSUE TITLE
Reconstruction error dominated by highly expressed genes?

### ISSUE THREAD

**Issue by vals** [YosefLab/scVI#297]
Hi,

My new years resolution for this year is to never work on log-counts as if they are continuous values. So when I do anything with single cell data, I fit models using e.g. negative binomial log likelihood. I have encountered a general issue I don't know how to deal with however, and I think scVI is affected by this too.

When using NB log likelihood as a loss function and fitting with several genes at once, genes with higher mean expression across all data will contribute more to loss when summing the results.

As an example, I looked at some scRNA-seq data using scVI. You can see the entire notebook here: https://nbviewer.jupyter.org/gist/vals/41ef66a55bafa5d10e477f617bc3a6d1 , but I'll summarise the main points below.

I made a `VAE`, and fitted it. Then I predicted all parameters for all data using the `VAE.inference` method.

I copied the code for negative binomial reconstruction error and made a version that returns the reconstruction error _per gene_.

```
def log_nb_positive_per_gene(x, mu, theta, eps=1e-8):
    """
    Note: All inputs should be torch Tensors
    log likelihood (scalar) of a minibatch according to a nb model.
    Variables:
    mu: mean of the negative binomial (has to be positive support) (shape: minibatch x genes)
    theta: inverse dispersion parameter (has to be positive support) (shape: minibatch x genes)
    eps: numerical stability constant
    """
    if theta.ndimension() == 1:
        theta = theta.view(1, theta.size(0))  # In this case, we reshape theta for broadcasting

    log_theta_mu_eps = torch.log(theta + mu + eps)

    res = theta * (torch.log(theta + eps) - log_theta_mu_eps) + \
        x * (torch.log(mu + eps) - log_theta_mu_eps) + \
        torch.lgamma(x + theta) - \
        torch.lgamma(theta) - \
        torch.lgamma(x + 1)
    
    return torch.sum(res, dim=0)
``` 

If I compare these reconstruction losses from the fitted model with the empirical mean of the counts in the expression matrix, genes with higher average counts have more loss.

![image](https://user-images.githubusercontent.com/668803/56327070-8720b100-612d-11e9-99ac-cac814d080bf.png)

My interpretation of this is that when the model is being fitted, latent variables and neural net parameters are going to optimize towards representing expression of the most highly expressed genes.

Any thoughts on this?

**Comment by romain-lopez** [comment]
Hi Vals, 

I guess the NB loss has more magnitude for most expressed genes. It's also a reason for which people would do for example whitening in machine learning as part of feature preprocessing ? That way the L2 loss weight each feature with the same contribution while otherwise the most variant feature is the most important ?

I guess we never looked into that for scVI development. Do you have a specific usecase in which that is a problem ?

Best, 
Romain

**Comment by romain-lopez** [comment]
Congratulations for your new year's resolution ! 🎖

**Comment by vals** [comment]
I encountered the issue in a supervised learning setting, where it caused issues because my model overfitted the most abundant gene.

But I wonder, if the most highly expressed genes contributes most to the loss, wouldn't that mean that most of the representation in the latent space will be used to model not-so-interesting housekeeping genes?

In general scVI finds good embeddings of data that often reproduce analysis from previous studies. So maybe it's not an issue? But one of the reasons to use count models instead of e.g. running PCA on the counts is to escape the mean-variance relation which implies that PCA will overfit to the few most highly expressed genes.

Perhaps it is the case that often the most highly expressed genes are what is important for cell types etc? But in that case simpler models should do fine?

It is sort of subtle, but I think there might be some conceptual odd thing that happens when this count model is used in a multivariate setting. Is there a reasonable way to do the equivalent of whitening for count models?

**Comment by jeff-regier** [comment]
Hi Valentine,

Thanks for the feedback. 

I'd expect highly expressed genes to matter somewhat more: we get to observe them more thoroughly so they are more informative. For rarely observed genes, we don't get a lot of new information from the data, so there's less basis for doubting/refining our prior knowledge (encoded in the prior distribution).

That said, the standard deviation of the negative binomial does scale with the mean, when theta (the inverse dispersion parameter) is held fixed. So we do expect greater variability for more highly expressed genes -- and our loss function penalizes deviations from the mean accordingly.

I don't recall whether we use a different setting of theta for each gene. (@romain-lopez Do you?) If some genes are affecting the latent representation too much, perhaps we could use a larger theta for those genes?

**Comment by jeff-regier** [comment]
Some discussion of how to estimate the inverse dispersion parameter here:
https://journals.sagepub.com/doi/pdf/10.3141/2019-03

If our current estimates of theta aren't good we may want to consider some alternative estimators.

**Comment by vals** [comment]
Hi Jeff,

Thanks for the link!

I do believe each gene get's it's own theta. You're saying maybe the theta can be scaled for "whitening"? In some way make the theta inversely proportional to the mean? This does sound kind of similar to e.g. how class imbalance is handled in classification.

But the mean-std relation is lower-bounded at std=mean right? Because in the limit case you get a Poisson. And it seems to me the same issue would be there if this was a multivariate Poisson likelihood?

In terms of mattering, maybe it's ambiguous. Certainly more of the data comes from the highly expressed genes, so in terms of globally explaining the data, making a model prioritize them will explain more of what you are observing from measurements.

But there are so many cases of expected interesting (biological) variation being due to more modestly abundant genes. If the latent representation is learned to explain variation in highly abundant this might miss the signal?  I did try to analyze one public dataset where I couldn't replicate the authors results, I'll have a look at that data again with this in mind. (Before I figured they had just over-processed the data).

--
Maybe a bit off topic for scVI: But would you have any ideas of how to deal with this in supervised learning? If I'm training a multitask (Gaussian) regression network I can whiten the data for each task and the least squares loss will be directly comparable between tasks. But in this setting a task with higher counts will systematically have higher loss.

**Comment by jeff-regier** [comment]
Indeed, the variance of a negative binomial is greater than or equal to its mean. By reducing theta (the inverse dispersion) for some of the highly expressed genes, you'd increase the variance modeled for those genes, thus reducing their influence on the latent space. That sounds like "whitening" .  I don't know if I'd recommend that as a solution though, because it seems rather ad hoc. I'd wonder why the original estimates of theta needed to be adjusted. I don't quite recall how we set theta or why we picked that particular estimator -- maybe it wasn't a good choice?

Another possibility is that theta is fine, but that you'd need more latent dimensions to model these interesting-but-less-prominent genes. We usually set the dimension of z to 8 I think, because it's not straightforward to get VAEs to effectively use many more dimensions than that.

**Comment by jeff-regier** [comment]
I'm not sure about the multitask regression network. Maybe the statistical model underlying it is misspecified for your dataset?

**Comment by jeff-regier** [comment]
Closed due to inactivity. Feel free to reopen this issue if there are further details.

**Comment by vals** [comment]
Yes that's fine. Thanks for the discussion!
